### PR TITLE
Update backup_lsn after each segment

### PR DESCRIPTION
**Pull request by petuhovskiy**
## Describe your changes

This helps when backup is interrupted by inconsistent election or anything else.

## Issue ticket number and link

Should help with https://github.com/neondatabase/neon/issues/3957, but will not fix it completely.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by arssher** [comment]
It will stuck on the next iteration waiting for changed commit_lsn. We need to either fix that or just do it more properly updating backup_lsn after each segment upload.

**Comment by petuhovskiy** [comment]
I also did the same, will push in a minute.

File: safekeeper/src/wal_backup.rs [hunk 1]
```diff
@@ -354,28 +349,35 @@ impl WalBackupTask {
...
         backup_single_segment(s, timeline_dir, workspace_dir)
             .await
             .with_context(|| format!("offloading segno {}", s.seg_no))?;
 
-        res = s.end_lsn;
+        let new_backup_lsn = s.end_lsn;
+        let res = timeline.set_wal_backup_lsn(new_backup_lsn);
+        if let Err(e) = res {
+            error!("failed to set wal_backup_lsn: {}", e);
+            return Err(e);
```
Comment (arssher) [inline safekeeper/src/wal_backup.rs:370]: It doesn't make sense to retry after canceled timeline, we can just terminate the whole task. OTOH, retrying is not  harmful, and AFAIR there is no adequate reaping of failed tasks, so can leave as is currently as well.

However, logging error in any case pointless as we do that above.

**Review by arssher — APPROVED** [review]

**Comment by arssher** [comment]
(feel free to take my commit message =))

File: safekeeper/src/wal_backup.rs (hunk 1 above)
Comment (petuhovskiy) [reply safekeeper/src/wal_backup.rs]: > It doesn't make sense to retry after canceled timeline

Yes, but only if we rely on the fact that `set_wal_backup_lsn` can return error only when timeline is cancelled.

> However, logging error in any case pointless as we do that above.

Replaced with context()

**Pull request merged**